File: src/boj/BOJ_02630_MakingColoredPaper.py
import sys


def divide_and_conquer(arr, row, col, size):
    global cnt_0, cnt_1
    isSame = True
    first_color = arr[row][col]
    for i in range(row, row + size):
        for j in range(col, col + size):
            if paper[i][j] != first_color:
                isSame = False

    if isSame:
        if first_color == 0:
            cnt_0 += 1
        else:
            cnt_1 += 1
    else:
        half_size = size // 2
        divide_and_conquer(paper, row, col, half_size)
        divide_and_conquer(paper, row, col + half_size, half_size)
        divide_and_conquer(paper, row + half_size, col, half_size)
        divide_and_conquer(paper, row + half_size, col + half_size, half_size)


N = int(sys.stdin.readline())
paper = []
cnt_0 = 0
cnt_1 = 0

for _ in range(N):
    row = list(map(int, sys.stdin.readline().split()))
    paper.append(row)

divide_and_conquer(paper, 0, 0, N)

print(cnt_0)
print(cnt_1)

File: src/boj/BOJ_02630_MakingColoredPaper.java
package boj;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class BOJ_02630_MakingColoredPaper {
    static int[][] paper;
    static int cnt_0, cnt_1;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int N = Integer.parseInt(br.readLine());
        paper = new int[N][N];
        cnt_0 = 0;
        cnt_1 = 0;

        for (int i = 0; i < N; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            for (int j = 0; j < N; j++) {
                paper[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        divideAndConquer(0, 0, N);

        System.out.println(cnt_0);
        System.out.println(cnt_1);
    }

    public static void divideAndConquer(int row, int col, int size) {
        boolean isSame = true;
        int firstColor = paper[row][col];
        for (int i = row; i < row + size; i++) {
            for (int j = col; j < col + size; j++) {
                if (paper[i][j] != firstColor) {
                    isSame = false;
                    break;
                }
            }
            if (!isSame) {
                break;
            }
        }

        if (isSame) {
            if (firstColor == 0) {
                cnt_0++;
            } else {
                cnt_1++;
            }
        } else {
            int halfSize = size / 2;
            divideAndConquer(row, col, halfSize);
            divideAndConquer(row, col + halfSize, halfSize);
            divideAndConquer(row + halfSize, col, halfSize);
            divideAndConquer(row + halfSize, col + halfSize, halfSize);
        }
    }
}
